### DISCUSSION TITLE
Installing Authentication Extensions gives errors

### DISCUSSION THREAD

**Discussion by ziomek64** [unoplatform/uno.extensions#2193] [Q&A] [answered]
Hello,
I'm trying to move from MAUI. Trying to do WebAuthentication on Android. When I installed extension for all projects in solution I get error UNOB0002 for DataContracts.dll. So can't proceed further.. What should I do?

Also docs mention to install Uno.Extensions.Authentication.WinUI . Why is It called WinUI at the end instead of installing just Uno.Extensions.Authentication? What If I go for mobile and material design?

I need help learning how to install stuff here. I'm targetting Android only. Went for Uno because it provides Material Design 3.0 support. Maui is far behind in that terms.

I'm trying to understand stuff in Uno, but it's so much less beginner-friendly, and I'm not pro dev. MS docs for MAUI are far more comprehensive.

**Comment by ziomek64** [comment]
So I managed to do that with some other template..
But can't get the browser to open when needed. I've set up default auth:
```
                .UseAuthentication(auth =>
    auth.AddWeb()
                )
```

Default LoginPage view:
```
<Page x:Class="UnoApp6.Presentation.LoginPage"
      xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
      xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
      xmlns:local="using:UnoApp6.Presentation"
      xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
      xmlns:uen="using:Uno.Extensions.Navigation.UI"
      xmlns:utu="using:Uno.Toolkit.UI"
      xmlns:um="using:Uno.Material"
      mc:Ignorable="d"
      NavigationCacheMode="Required"
      Background="{ThemeResource BackgroundBrush}">

  <Grid utu:SafeArea.Insets="VisibleBounds">
    <Grid.RowDefinitions>
      <RowDefinition Height="Auto" />
      <RowDefinition />
    </Grid.RowDefinitions>
    <utu:NavigationBar Content="{Binding Title}" />

    <StackPanel Grid.Row="1"
          HorizontalAlignment="Center"
          VerticalAlignment="Center"
          Width="200"
          Spacing="16">
      <Button Content="Login"
          Command="{Binding Login}"
          HorizontalAlignment="Stretch" />
    </StackPanel>
  </Grid>
</Page>
```

Default viewmodel for that:
```
namespace UnoApp6.Presentation;

public partial class LoginViewModel : ObservableObject
{
    private IAuthenticationService _authService;

    private INavigator _navigator;

    private IDispatcher _dispatcher;


    public LoginViewModel(
        IDispatcher dispatcher,
        INavigator navigator,
        IAuthenticationService authService)
    {
        _dispatcher = dispatcher;
        _navigator = navigator;
        _authService = authService;
        Login = new AsyncRelayCommand(DoLogin);
    }

    private async Task DoLogin()
    {
        var success = await _authService.LoginAsync();
        Console.WriteLine("Login");
    }

    public string Title { get; } = "Login";

    public ICommand Login { get; }
}
```

It definitely goes inside the task as it prints Login. My appsettings.json:
```
{
  "AppConfig": {
    "Environment": "Production"
  },
  "Web": {
    "LoginStartUri": "https://anilist.co/api/v2/oauth/authorize?client_id=15408&response_type=token",
    "LogoutStartUri": "https://anilist.co/api/v2/oauth/authorize?client_id=15408&response_type=token"
  }
}
```

**Reply by nickrandolph** [reply]
Are you able to share a sample project?

**Reply by ziomek64** [reply]
@nickrandolph https://github.com/ziomek64/UnoApp6
This is literally just template, with my login uris etc. Here you go

**Reply by ziomek64** [reply]
> Are you able to share a sample project? 

Any ideas?

**Comment by nickrandolph** [comment]
> Hello, I'm trying to move from MAUI. Trying to do WebAuthentication on Android. When I installed extension for all projects in solution I get error UNOB0002 for DataContracts.dll. So can't proceed further.. What should I do?
> 
> Also docs mention to install Uno.Extensions.Authentication.WinUI . Why is It called WinUI at the end instead of installing just Uno.Extensions.Authentication? What If I go for mobile and material design?
> 
> I need help learning how to install stuff here. I'm targetting Android only. Went for Uno because it provides Material Design 3.0 support. Maui is far behind in that terms.
> 
> I'm trying to understand stuff in Uno, but it's so much less beginner-friendly, and I'm not pro dev. MS docs for MAUI are far more comprehensive.

Uno.Extensions.Authentication.WinUI should only be installed into the main class library, not into the DataContract library. Although it's got the WinUI suffix, it's compatible with all Uno targets. The WinUI suffix is used to differentiate from the UWP based libraries that Uno offers

**Reply by ziomek64** [reply]
Will I get any help here?

**Reply by nickrandolph** [reply]
Hey @ziomek64 apologies for the delay. Let me take a look at this and see what I can find out

**Comment by nickrandolph** [comment]
@ziomek64 this PR (https://github.com/unoplatform/uno.extensions/pull/2198/) should address this issue. Essentially what was happening is that the configuration information wasn't being picked up, so the loginuri was null, which meant the browser wasn't being launched.

**Reply by ziomek64** [reply]
@nickrandolph  How often do dev versions of nugets get released ? Or would I have to compile it myself most likely

**Comment by ziomek64** [comment]
@nickrandolph It's just not working for me, I don't know what's wrong to be honest. Browser doesn't open. I think I upgraded packages correctly (to be honest uno packages is a mess for me coming from maui)

https://github.com/ziomek64/UnoApp6/

**Comment by nickrandolph** [comment]
@ziomek64 the challenge we have is that we have a much wider offering that maui, meaning that package management has been an issue. With your sample application I would suggest upgrading all packages to the latest stable release, and then only update extensions to the latest dev package.

**Comment by nickrandolph** [comment, marked as answer]
I've tried your sample and the issue is now that you're not specifying either a redirect uri as part of the LoginStartUri (eg https://github.com/unoplatform/uno.extensions/blob/978e173046c23b221712f7ceba40aab7928ec1ee/testing/TestHarness/TestHarness.Shared/Ext/Authentication/Web/appsettings.webauthsettings.json#L3), or providing the LoginCallbackUri parameter. If you provide either of these you should see the browser being displayed.
I'll endeavour to add more detailed logging to the auth process to make it easier to determine what's missing in future

**Reply by ziomek64** [reply]
@nickrandolph  Seems to be working. I had trouble doing it in LoginStartUri, i've set up LoginCallbackUri parameter in app.cs as you told me. How I can easily access the token at all times in a view model, in my app in general? I need to use it for API requests.

**Reply by nickrandolph** [reply]
If you register the endpoint using Uno
Extensions.http the token should be automatically applied. Otherwise you can take a dependency in your viewmodel on ITokenCache and retrieve the tokens directly
